rpc_util_test.go

```diff
@@ -290,3 +292,130 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
 func BenchmarkGZIPCompressor1MiB(b *testing.B) {
 	bmCompressor(b, 1024*1024, NewGZIPCompressor())
 }
+
+type mockCompressor struct {
+	DecompressedData []byte
+	ErrDecompress    error
+	Size             int
+	CustomReader     io.Reader
+}
+
+func (m *mockCompressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	return nil, nil
+}
+
+func (m *mockCompressor) Decompress(r io.Reader) (io.Reader, error) {
+	if m.ErrDecompress != nil {
+		return nil, m.ErrDecompress
+	}
+	if m.CustomReader != nil {
+		return m.CustomReader, nil
+	}
+	return bytes.NewReader(m.DecompressedData), nil
+}
+
+func (m *mockCompressor) DecompressedSize(compressedBytes mem.BufferSlice) int {
+	return m.Size
+}
+
+func (m *mockCompressor) Name() string {
+	return "mockCompressor"
+}
+
+type ErrorReader struct{}
+
+func (e *ErrorReader) Read(p []byte) (n int, err error) {
+	return 0, errors.New("simulated io.Copy read error")
+}
+
+func TestDecompress(t *testing.T) {
+	tests := []struct {
+		name                  string
+		compressor            encoding.Compressor
+		input                 mem.BufferSlice
+		maxReceiveMessageSize int
+		want                  mem.BufferSlice
+		size                  int
+		error                 error
+	}{
+		{
+			name: "Successful decompression",
+			compressor: &mockCompressor{
```

Comment:
do we really need to test compressor for all cases? Can't we just use a regular compressor?

Reply:
Part of this test is to also verify that inputs work with regular compressor

Reply:
Something like this @purnesh42H 

``compressor := gzip.NewCompressor() 
	originalData := []byte("test data")

	// Compress
	compressedData, err := compressor.Compress(originalData)
	if err != nil {
		t.Fatalf("Failed to compress data: %v", err)
	}

	// Decompress
	decompressedData, err := compressor.Decompress(compressedData)
	if err != nil {
		t.Fatalf("Failed to decompress data: %v", err)
	}

	if !bytes.Equal(decompressedData, originalData) {
		t.Errorf("Decompressed data does not match original")
	}``

Reply:
yeah i think we should try to get the regular test cases working with regular compressor. Only use testCompressor when we want to simulate or enforce an error

Reply:
can you check this and see how are different compressors are tested https://github.com/grpc/grpc-go/blob/56df169480cdb4928a24a50b5289f909f0d81ba7/test/compressor_test.go ?

We should try to do something similar. There are test compressors as well regular compressors based on test cases. Compressor can be part of test case struct

Reply:
One more example https://github.com/grpc/grpc-go/blob/56df169480cdb4928a24a50b5289f909f0d81ba7/rpc_util_test.go#L151